content/posts/lets-code-contact-form-in-phoenix-part-two.md

```diff
@@ -0,0 +1,264 @@
+
+```elixir
+# test/my_app/support_test.exs
+
+defmodule MyApp.SupportTest do
+  # The `DataCase` module ships with a standard Phoenix app.
+  # See the docs for more info:
+  # https://hexdocs.pm/phoenix/testing_contexts.html#the-datacase
+  #
+  # The `async: true` option allows the tests in this file to run concurrently
+  # with other tests in the application. Tests in this file still run serially.
+  use MyApp.DataCase, async: true
+
+  alias MyApp.Support
+
+  test "changing a message" do
+    # `Support.change_message/1` is expected to return a Changeset for a Message.
+    assert %Ecto.Changeset{data: %Support.Message{}} = Support.change_message()
+  end
+end
+```
+
+Running this test results in the following failure:
+
+```
+** (UndefinedFunctionError) function MyApp.Support.change_message/0 is undefined (module MyApp.Support is not available)`
+```
+
+This is stating that not only are we missing the `change_message` function, but the `MyApp.Support` module doesn't even exist. Let's fix that by creating the `Support` module and `change_message` function:
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule MyApp.Support do
+  @moduledoc """
+  The Support module is the interface to customer support for the application.
+  Look to this module when you want to provide support to application users.
+  """
+
+  # Alias the `Support.Message` module for easy reference later in this module.
+  alias __MODULE__.Message
+
+  @doc """
+  Returns a changeset for a `Support.Message`.
+  """
+  @spec change_message() :: Ecto.Changeset.t()
+  def change_message do
+    %Message{}
+    |> Message.changeset(%{})
+  end
+end
+```
+
+This very simple module and `change_message` function is enough to make the test pass. This function can be used to build a changeset for a new message. Now I will add tests for sending a message.
+
+```elixir
+# test/my_app/support_test.exs
+
+defmodule MyApp.SupportTest do
+  describe "sending a message" do
+    test "with valid fields" do
+      fields = %{
+        email: "barry@bluejeans.test",
+        subject: "Halp Me",
+        body: "I am having problems."
+      }
+
+      assert {:ok, %Support.Message{} = message} = Support.send_message(fields)
+      assert message.email == "barry@bluejeans.test"
+      assert message.subject == "Halp Me"
+      assert message.body == "I am having problems."
+    end
+
+    test "with invalid fields" do
+      assert {:error, cset} = Support.send_message(%{})
+
+      refute cset.valid?
+      assert "can't be blank" in errors_on(cset).email
+    end
+  end
+end
+```
+
+Here, I added a new `describe` block for testing the functionality of sending a message. This block has two tests: one for a successful message send and one for a failed message send. Successfully sending a message results in an `:ok` tuple response with a sent message. Failing to send a message returns an `:error` tuple with a changeset.
```

Comment:
Maybe it'd be nice to call out that this is a pretty standard way for functions to behave in elixir, and link to some docs that describe that pattern?